### DISCUSSION TITLE
Merging related objects of two objects

### DISCUSSION THREAD

**Discussion by ljiatu** [sqlalchemy/sqlalchemy#6530] [Usage Questions] [answered]
Suppose we have an ORM model `A` with a bunch of relationships, `A.bs,` `A.cs` etc. where `B`, `C`, ... are models themselves. Let's say we have two instances of `A`, `a1` and `a2`, stored in the DB, and each of them is associated with their own collections of `B`s and `C`s. Let's say now that we need to "collapse" `a1` and `a2` by deleting `a2` and merge all of `a2`'s related objects into `a1`.

We previously used the `merge_references` function in `sqlalchemy_utils`: https://sqlalchemy-utils.readthedocs.io/en/latest/foreign_key_helpers.html?highlight=merge_references#merge-references, but then we got rid of the dependency on `sqlalchemy_utils` since it wasn't compatible with SQLAlchemy 1.4 at the time we migrated to 1.4. So currently we are manually enumerating all relationships of `A` and updating the fkeys 1-by-1. This is clearly not very scalable and it's easy to miss if we add more relationships to `A`. Is this achievable with `session.merge()` somehow? From reading the docs, I'm getting the impression that `merge` is mainly for merging an object that's outside of a session with something that's already in the session, so not sure if this approach makes sense.

Could someone advise the best way to go about this? Your help would be much appreciated! 🙏

**Comment by zzzeek** [comment, marked as answer]
if sqlalchemy-utils did exactly what you needed here, then it seems like that's the function you need!   why not keep using it?   session.merge() sort of does this in that it will add things to an identity but it doesnt remove things from another identity.

**Reply by ljiatu** [reply]
Because:
1. It broke at the time we migrated to SQLAlchemy 1.4, so we moved off of it to unblock the migration
2. There is no async support. Not sure if there will ever be, and if so, when. The pacakge doesn't appear to be very actively maintained.
3. Part of the reason why it broke was that it depends on internal implementation details of SQLAlchemy, e.g. importing "private" members into a module. So I'm concerned that it'll break again.

It's a great package that helped us a lot in the past, but I just don't feel like it's going to be worth the maintenance effort for us to keep using it for just a few functions.

**Reply by zzzeek** [reply]
> Because:
> 
>     1. It broke at the time we migrated to SQLAlchemy 1.4, so we moved off of it to unblock the migration

they've had new releases and all their problems should be fixed now.  You can always just [vendor and/or port](https://github.com/kvesteri/sqlalchemy-utils/blob/a4154bd0809bc6bbf0c27d5f7c0f3f2872edd779/sqlalchemy_utils/functions/foreign_keys.py#L100) the routine there it looks pretty simple.

> 
>     2. There is no async support. Not sure if there will ever be, and if so, when. The pacakge doesn't appear to be very actively maintained.

again I'd vendor the function and/or send them a PR.


> 
>     3. Part of the reason why it broke was that it depends on internal implementation details of SQLAlchemy, e.g. importing "private" members into a module. So I'm concerned that it'll break again.

hopefully they've fixed that.

> 
> 
> It's a great package that helped us a lot in the past, but I just don't feel like it's going to be worth the maintenance effort for us to keep using it for just a few functions.

OK just from my POV, you are asking us (sqlalchemy) to re-invent a new set of code for a use case that someone has gone through the effort to write the code to solve your problem exactly.   the code given is open source so I would suggest just reading it and adapting it to your use case.

**Reply by ljiatu** [reply]
Thanks for the suggestion, and no that's not what I'm asking at all. You were asking why we would not just keep using sqlalchemy-utils, and I was just laying out exactly why we wouldn't want to use it. In my original question, I was also just asking if anything that already exists in SQLAlchemy could help solve our problem out-of-the-box. Never once had I ever asked SQLAlchemy to re-invent the wheels, neither would I ever do that just for my own specific usecase. With all due respect, I don't think it's super productive to make the assumptions that I'm asking something outrageous that I never intended to.

Thank you again for your suggestions and the great library you guys put together. Hope you have a great day.

**Reply by zzzeek** [reply]
OK I apologize for making that leap, but when you asked "is there a way for merge() to do this", the answer is "probably, with some additonal object manipulation code around it".   If sqlalchemy-utils didn't exist, I would then be thinking about how to write a generalized function for you that would iterate through relationships and all that.   So I don't see how I could answer the question "is there a way to make merge() do this" without my coming up with a recipe of some kind.  

again sorry that hit you in a bad way but I do this all day and if you know me, you know that when someone needs to do something, i pretty much write out complete examples for them, and Ive wanted other projects like sqlalchemy-utils to be picking up some of that off of me.    I hope this works out for you.

**Reply by ljiatu** [reply]
No worries. Glad that we were able to clear up the miscommunication. Thanks again for your suggestion!
